### PR TITLE
Init test structure

### PR DESCRIPTION
**Pull request by PooyaHekmati**
There was a test folder under RUFAS which used unittest. The content of this folder was inaccurate and did not follow clean code principles. Removed this folder. Added a new folder called tests next to RUFAS (where it should be), this folder contains files for tests, each corresponding to a module. Files that do not belong to any module (those who are directly under RUFAS) are tested in misc.py. Also, used pytest instead of unittest to resolve the issues we had with unittest for relative importing. Pytest is also preferred for CI pipelines. To run tests, run "pytest <filename>" in tests folder e.g., to run test_animal.py, run pytest test_animal.py.

### CODE REVIEW

**Comment by KFosterReed** [comment]
This pull request looks good to me. All the test files run as expected since the tests are empty but still a good sign and starting point! I like that there is a separate test file for the output handler because those files have been a bit cumbersome to update as the simulation model updates so this should help with those updates as well.

**Review by KFosterReed — APPROVED** [review]

File: tests/test_feed.py
Comment (militsasotirova) [inline tests/test_feed.py:4]: Should the 'description' in the comment at the top of each file state the module being tested? For instance here it would be "Implements test cases for the feed module." or something similar.

File: tests/test_feed.py
Comment (militsasotirova) [inline tests/test_feed.py:17]: A general question that applies to the rest of the repo as well (maybe something for our fix-a-thon?) - moving forward, are we sticking to a character per line limit? I think at one point a few years ago we mentioned keeping lines at 80 characters but I don't think we really enforced that. This has me thinking that it would be good to have RuFaS code style guidelines (e.g. style for comments, character per line limits, alphabetizing import statements, etc.) that we ensure are followed in each PR. We don't necessarily have to address this in this PR, just something I've been thinking about.

File: tests/test_feed.py
Comment (militsasotirova) [inline tests/test_feed.py:12]: While the code runs without errors, I do get warnings about the file paths in these comments. (I only get them the first time I run the command, I'm guessing the executable part of the `pytest` code is cached and the part that checks for these warnings is only run when the code itself is changed. In any case, it's a simple fix.) 
```
DeprecationWarning: invalid escape sequence \c
    """Unit test for function update_all in file routines\feed\carbon_loss.py"""
```
This comes from the fact that you have a '\\' instead of a '/' in the file path and the string is looking to apply the escape character '\c' which doesn't exist. If we changed all instances of '\\' to '/' we wouldn't get these warnings.

File: tests/test_output_handler.py
Comment (militsasotirova) [inline tests/test_output_handler.py:131]: Some of the function names in this file are repeated. They're for functions in different files, but we should rename them to be more specific to the file they are in.

File: tests/test_manure_storage.py
Comment (militsasotirova) [inline tests/test_manure_storage.py:26]: There are also a few repeated function names in this file.

File: tests/test_field.py
Comment (militsasotirova) [inline tests/test_field.py:96]: More repeated function names in this file. Also in` test_animal.py` and `test_feed.py`.

**Comment by militsasotirova** [comment]
Everything generally looks good - I’ve made a few comments above (I'll approve once we resolve those). I’m able to run all of the tests without errors.

What are your thoughts on adding functionality to run all the unit tests at once? I’m thinking a bash script that runs ‘pytest test_*.py’ for all of the files in the `tests` directory or something, keeping track of how many tests have passed out of how many total and presenting that information to the user at the end.

EDIT: I was doing some research on the purest framework and if you simply enter `pytest` into the command line within the `tests` directory, all of the test files will be run.

Also a note for the future: I agree that the animal ration and manure tests are likely out of date so we should start over with those - but whoever populates the field tests could probably use what Max wrote in the files that have been deleted in this PR as a starting point.

File: tests/test_feed.py
Comment (PooyaHekmati) [reply tests/test_feed.py:4]: By convention, the name of the unit which is under test is mentioned on the file name after "test_" prefix. Generally speaking, we would like to avoid having redundant information as it can lead to inconsistency in the future. Moreover, the structure of RUFAS is about to change, so we don't want to tie up our test structure to code structure.

File: tests/test_feed.py
Comment (PooyaHekmati) [reply tests/test_feed.py:17]: The style guideline you mentioned can be handled using automatic pipelines (linting). As mentioned in the EOY announcement, we will be considering both linting and PEP in the near future; but we are not talking about them right now to let the team focus on things with higher priorities (unit testing and docstring). We will apply automatic linting, so I am not concerned about that right now.

File: tests/test_feed.py
Comment (PooyaHekmati) [reply tests/test_feed.py:12]: This will be fixed in the linting process.

File: tests/test_output_handler.py
Comment (PooyaHekmati) [reply tests/test_output_handler.py:131]: This is an indication of (1) bad design, (2) not practicing DRY (3) not practicing SRP (4) not having reasonable names. Instead of using painkillers and bypassing the issue by renaming test cases (which is against the convention so it makes the code dirty and has the potential for leaving holes in the test suite), we need to fix the problem in the root. By convention, each test unit should be named "test_<unit_name>".

**Comment by PooyaHekmati** [comment]
I am not sure about having a bash file for running the test suite, it is not a common practice. The reason is that when developing, we just want to run test cases that are relevant, so we put them in separate files. Running the entire test suite is computationally expensive, we only do it once a major action such as merging happens. We don't need to worry about it though, because we will have automated pipelines doing it and reporting the results.

Regarding the deleted test cases, I deleted them because (1) they were in the wrong place, (2) not having a code is much better than having a code that is wrong, (3) there were noticeable problems in those tests, (4) oftentimes developing something from the scratch saves resources compared to fixing existing code, (5) they will be always accessible in git's history, so they are not gone (6) we want our programmers to understand and verify the code, having them to develop test cases is a powerful tool to achieve this.

File: tests/test_output_handler.py
Comment (militsasotirova) [reply tests/test_output_handler.py:131]: How should overridden methods be dealt with in this case?

Comment (PooyaHekmati) [reply tests/test_output_handler.py:131]: We have to develop at least 544 unit tests, most of which have no problem. We will focus on those first and by the time we get to these problematic ones, the problem is already solved as we have reworked the code structure by then. In case someone insists on working on one of these test cases, they have a few options: (1) fix the name conflict in the code (2) refactor the code and employ DRY, (3) refactor the code and employ SRP, (4) refactor and employ inheritance. The bottom line is that we cannot change the test suite because the codebase has a problem. The whole purpose of having tests is to identify and fix the problem, not simply develop something that agrees with the code. A good test is a test that fails, a passing test provides no information and there is no point in investing resources in it.

**Review by militsasotirova — APPROVED** [review]

**Pull request merged**
